### ISSUE TITLE
Question: MaskedCoupling vs SplitCoupling

### ISSUE THREAD

**Issue by oliverdutton** [deepmind/distrax#132]
I see in the example of `flow.py` you use masked coupling which you split per layer.

Am I correct in guessing split coupling when used naively on a 1D, even length event shape (eg (2,)) like below is equivalent to the masked coupling 

Split way of looking at it:
```python
  for i in range(num_layers):
    layer = distrax.SplitCoupling(
        split_index=ndim//2,
        event_ndims=1,
        bijector=bijector_fn,
        conditioner=make_conditioner([ndim//2], hidden_sizes,
                                    num_bijector_params),
        swap=(i%2)==1
    )
    layers.append(layer)
```

Mask way of looking at it
```python
  for _ in range(num_layers):
    layer = distrax.MaskedCoupling(
        mask=mask,
        bijector=bijector_fn,
        conditioner=make_conditioner(event_shape, hidden_sizes,
                                     num_bijector_params))
    layers.append(layer)
    # Flip the mask after each layer.
    mask = jnp.logical_not(mask)
```

With the difference being that the masked coupling is more general (ie higher dim event shapes) and explicit control of which dimensions interact.

Is there a performance penalty for the masked coupling, because it looks like it makes the MLP with double the input size and double the output. Though realistically all the computational cost will be in those hidden layers which are invariant to the event shape. The fact your conditioner basically just gets a load of zero's in the input doesn't matter much.

**Comment by gpapamak** [comment]
Yes, you're right. `SplitCoupling` and `MaskedCoupling` are nearly equivalent, they are just different ways of implementing the same concept. We provide both of them for convenience: depending on what you want to do, one may be more convenient to use than the other.

`SplitCoupling` is more convenient when you want to partition the input along some axis, `MaskedCoupling` is more convenient when you want to use arbitrary partitions. You can still use `SplitCoupling` with arbitrary partitions though: you first permute input dimensions so that the part you want to transform comes first, and then you undo the permutation in the output. One small difference is that `SplitCoupling` can use arbitrary inner bijectors, whereas `MaskedCoupling` has to use an elementwise inner bijector (because the mask can be arbitrary, the inner bijector must be constrained to operate independently on each maskable element).

Regarding performance, `MaskedCoupling` does more work than needed because the masked-out inputs and outputs of the inner bijector are not used, so it may end up using more memory than an equivalent `SplitCoupling`. But the masking is very parallelizable, so depending on the accelerator you use you may not see a runtime penalty.

**Comment by oliverdutton** [comment]
Thank you for the thorough answer. I just wanted to be sure.

Side note: I love the package, thank you.
